File: python/tests/smoke.py
"""실서버 스모크 테스트 (수동 실행 전용).

    python tests/smoke.py next     # NEXT_CLIENT_ID / NEXT_CLIENT_SECRET
    python tests/smoke.py kis      # KIS_APPKEY / KIS_APPSECRET / KIS_CANO
    python tests/smoke.py kiwoom   # KIWOOM_APPKEY / KIWOOM_SECRETKEY

시세 -> 캔들 -> 캘린더 -> 계좌 -> 보유 -> 주문가능액 -> (장중이면) 주문 생성->조회->취소.
"""
import logging
import os
import sys
from decimal import Decimal

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
logging.basicConfig(level=logging.INFO, format="%(levelname)s %(name)s %(message)s")

from hermetix import (  # noqa: E402
    CandleInterval, CreateOrderRequest, KisClient, KiwoomClient, NextClient,
    MarketClosedError, BrokerApiError, OrderSide, OrderType,
)
from hermetix.engine import MarketCalendar, pnl_report  # noqa: E402


def make_client(name: str):
    if name == "next":
        return NextClient(os.environ["NEXT_CLIENT_ID"], os.environ["NEXT_CLIENT_SECRET"]), "AAPL", Decimal("150")
    if name == "kis":
        return KisClient(os.environ["KIS_APPKEY"], os.environ["KIS_APPSECRET"], os.environ["KIS_CANO"]), "005930", None
    if name == "kiwoom":
        return KiwoomClient(os.environ["KIWOOM_APPKEY"], os.environ["KIWOOM_SECRETKEY"]), "005930", None
    raise SystemExit(f"unknown broker: {name}")


def main(name: str) -> None:
    client, symbol, far_price = make_client(name)
    print(f"== {name} ({client.capabilities.market}) ==")

    quotes = client.get_quotes([symbol])
    assert quotes and quotes[0].price > 0
    print(f"현재가 {symbol}: {quotes[0].price} / 등락률 {quotes[0].change_rate}")

    candles = client.get_candles(symbol, CandleInterval.DAY_1, 30)
    assert len(candles) >= 20 and candles[0].timestamp < candles[-1].timestamp
    print(f"일봉 {len(candles)}개 / 최신 종가 {candles[-1].close}")

    print(f"정규장 open = {MarketCalendar(client).is_regular_open()}")

    account = client.get_account()
    assert account.cash > 0
    print(f"예수금 {account.cash} / 총평가 {account.portfolio_value}")

    holdings = client.get_holdings()
    print(f"보유 {len(holdings)}종목: {[(h.symbol, str(h.quantity)) for h in holdings[:5]]}")

    power = client.get_buying_power()
    assert power > 0
    print(f"주문가능 {power}")

    client.get_orders()
    client.get_fills()

    # 주문 사이클 - 시장가와 먼 지정가 매수 후 취소
    if far_price is None:
        far_price = (quotes[0].price * Decimal("0.8")).quantize(Decimal("100"))
    try:
        order = client.create_order(CreateOrderRequest(
            symbol=symbol, side=OrderSide.BUY, order_type=OrderType.LIMIT,
            quantity=Decimal(1), limit_price=far_price))
        print(f"주문 접수: {order.order_id}")
        detail = client.get_order(order.order_id)
        print(f"주문 조회: {detail.status.value}")
        canceled = client.cancel_order(order.order_id)
        print(f"주문 취소: {canceled.status.value}")
    except (MarketClosedError, BrokerApiError) as e:
        print(f"주문 스킵: {e.message}")

    report = pnl_report(client)
    print(f"PnL: portfolio={report['portfolio_value']} unrealized={report['total_unrealized_pnl']}")
    print("SMOKE OK")


if __name__ == "__main__":
    main(sys.argv[1] if len(sys.argv) > 1 else "next")

File: js/tests/smoke.ts
/**
 * 실서버 스모크 (수동 실행 전용):
 *   node dist/tests/smoke.js next|kis|kiwoom   (환경변수로 키 주입)
 */
import { Decimal } from "decimal.js";
import { BrokerApiError, KisClient, KiwoomClient, MarketCalendar, NextClient, pnlReport } from "../src/index.js";
import type { BrokerClient } from "../src/index.js";

function makeClient(name: string): [BrokerClient, string, Decimal | null] {
  const env = process.env;
  if (name === "next") return [new NextClient(env.NEXT_CLIENT_ID!, env.NEXT_CLIENT_SECRET!), "AAPL", new Decimal(150)];
  if (name === "kis") return [new KisClient(env.KIS_APPKEY!, env.KIS_APPSECRET!, env.KIS_CANO!), "005930", null];
  if (name === "kiwoom") return [new KiwoomClient(env.KIWOOM_APPKEY!, env.KIWOOM_SECRETKEY!), "005930", null];
  throw new Error(`unknown broker: ${name}`);
}

async function main(name: string): Promise<void> {
  const [client, symbol, farPriceFixed] = makeClient(name);
  console.log(`== ${name} (${client.capabilities.market}) ==`);

  const [quote] = await client.getQuotes([symbol]);
  if (!quote.price.gt(0)) throw new Error("price <= 0");
  console.log(`현재가 ${symbol}: ${quote.price} / 등락률 ${quote.changeRate}`);

  const candles = await client.getCandles(symbol, "1d", 30);
  if (candles.length < 20 || candles[0].timestamp >= candles[candles.length - 1].timestamp) throw new Error("candles bad");
  console.log(`일봉 ${candles.length}개 / 최신 종가 ${candles[candles.length - 1].close}`);

  console.log(`정규장 open = ${await new MarketCalendar(client).isRegularOpen()}`);

  const account = await client.getAccount();
  if (!account.cash.gt(0)) throw new Error("cash <= 0");
  console.log(`예수금 ${account.cash} / 총평가 ${account.portfolioValue}`);

  const holdings = await client.getHoldings();
  console.log(`보유 ${holdings.length}종목: ${holdings.slice(0, 5).map((h) => `${h.symbol}x${h.quantity}`).join(", ")}`);

  const power = await client.getBuyingPower();
  if (!power.gt(0)) throw new Error("buying power <= 0");
  console.log(`주문가능 ${power}`);

  await client.getOrders();
  await client.getFills();

  const farPrice = farPriceFixed ?? quote.price.mul("0.8");
  try {
    const order = await client.createOrder({
      symbol, side: "BUY", orderType: "LIMIT", quantity: new Decimal(1), limitPrice: farPrice,
    });
    console.log(`주문 접수: ${order.orderId}`);
    console.log(`주문 조회: ${(await client.getOrder(order.orderId)).status}`);
    console.log(`주문 취소: ${(await client.cancelOrder(order.orderId)).status}`);
  } catch (e) {
    if (e instanceof BrokerApiError) console.log(`주문 스킵: ${e.message}`);
    else throw e;
  }

  const report = await pnlReport(client);
  console.log(`PnL: portfolio=${report.portfolioValue} unrealized=${report.totalUnrealizedPnl}`);
  console.log("SMOKE OK");
}

main(process.argv[2] ?? "next").catch((e) => {
  console.error("SMOKE FAILED:", e);
  process.exit(1);
});
